<?xml version="1.0"?>
<launch>
    <!-- RealSense Camera Node -->
    <node
        pkg="realsense2_camera"
        exec="realsense2_camera_node"
        name="camera"
        output="screen">
        <param name="serial_no" value="" />
        <param name="enable_pointcloud" value="true"/>
        <param name="enable_pointcloud" value="true" />
        <param name="pointcloud.enable" value="true" />
        <param name="pointcloud.stream_filter" value="2" />
        <param name="pointcloud.stream_format_filter" value="XYZRGB" />
        <param name="align_depth.enable" value="true" />
        <param name="colorizer.enable" value="true" />
        <param name="publish_tf" value="true" />
        <param name="tf_publish_rate" value="0.0" />
    </node>

    <!-- White Pixel Filter Node -->
    <node
        pkg="perception_pipeline"
        exec="white_filter"
        name="white_filter"
        output="screen">
        <param name="input_topic" value="/camera/camera/depth/color/points" />
        <param name="output_topic" value="/filtered_points" />
        <param name="hsv_lower" value="0,0,200" />
        <param name="hsv_upper" value="180,30,255" />
        <param name="min_intensity" value="200" />
        <param name="min_brightness" value="180" />
        <param name="max_color_variation" value="30" />
        <param name="min_z" value="-0.1" />
        <param name="max_z" value="0.5" />
        <param name="max_distance" value="3.0" />
    </node>

    <!-- Costmap Generator Node -->
    <node
        pkg="perception_pipeline"
        exec="costmap_generator"
        name="costmap_generator"
        output="screen">
        <param name="input_topic" value="/filtered_points" />
        <param name="output_topic" value="/costmap" />
        <param name="resolution" value="0.05" />
        <param name="width" value="10.0" />
        <param name="height" value="10.0" />
        <param name="inflation_radius" value="0.3" />
    </node>

    <!-- Visualizer Node -->
    <node
        pkg="perception_pipeline"
        exec="visualizer"
        name="perception_visualizer"
        output="screen">
        <param name="costmap_topic" value="/costmap" />
        <param name="filtered_points_topic" value="/filtered_points" />
    </node>

    <!-- TF Static Transforms -->
    <node
        pkg="tf2_ros"
        exec="static_transform_publisher"
        name="camera_tf"
        args="0 0 0 0 0 0 base_link camera_link" />
    
    <node
        pkg="tf2_ros"
        exec="static_transform_publisher"
        name="map_tf"
        args="0 0 0 0 0 0 map base_link" />

    <!-- RViz for visualization -->
    <node
        pkg="rviz2"
        exec="rviz2"
        name="rviz"
        args="-d $(find-pkg-share perception_pipeline)/config/perception.rviz"
        output="screen" />
</launch>